src/qibolab/pulses.py

```diff
@@ -218,7 +218,7 @@ def eval(value: str) -> "PulseShape":
         shape_name = re.findall(r"(\w+)", value)[0]
         if shape_name not in globals():
             raise ValueError(f"shape {value} not found")
-        shape_parameters = re.findall(r"[\w+\d\.\d]+", value)[1:]
+        shape_parameters = re.findall(r"[-\w+\d\.\d]+", value)[1:]
```

Comment:
_Premise: this is in `eval` that is already replaced in 0.2, so if it's passing, I wouldn't care that much._

`[abcd]` matches any character within (once), and `[abcd]+` matches one or more of occurrences of any characters in the list.
Thus having twice `\d` (a digit) makes no sense.

I assume `[...]` has been used in place of `(...)`, that is just a group (so the elements inside have to be matched in the order they are written).

`\w` is a word character (i.e. a digit or letter or `_`), and I don't see how you need that for the parameters (unless you were writing them as kwargs, but then you'd need also an `=`), so it's essentially matching just digits, and we're lucky that `[...]` is simply making it ignored (since `\d` twice and `\w` just means `\w`, inside `[]`, and it's going to match only digits).

Reply:
I believe that the pattern was meant to match numbers, and not `,`, that are used as a separator.

Thus the pattern should have been:
```py
r"[+-]?\d+(\.\d*)?([eE][+-]?\d+)?"
```
I.e.:
- either `+` or `-` or none of them (but not both)
- then one or more digits
- possibly a `.` followed by zero or more digits
- possibly a group, led by an `e` (upper or lower case), with either `+` or `-` or none of them (but not both) and then one or more digits

I didn't test it, but if you want to confirm you can look for a float pattern, or just test yourself.

Reply:
However, since the idea is just parsing comma-separated float (or integers, that are float compatible), I'd suggest just the following:

```py
shape_parameters = [float(p) for p in value.split(",")]
```

Unless there are also booleans involved (for which you should try one or the other).

In any case, I'd convert to a programmatic parse (leveraging standard utilities) rather than using regexes...

Reply:
Thanks for the summary @alecandido. I actually learnt several things from these comments (which I could most likely also find online, but thanks for summarizing). Generally the idea behind this PR was exactly

> _Premise: this is in `eval` that is already replaced in 0.2, so if it's passing, I wouldn't care that much._

so I did a fix that required the minimal effort (and time) and since both pytests and the hardware check seemed to work, I didn't care much more.

Also, note
https://github.com/qiboteam/qibolab/blob/c6945e6c3237265d83c80a6f959898f4b44a30fa/src/qibolab/pulses.py#L222
which probably means that it is known that the current approach is not robust. Also, none of these approaches will work for the `Custom` shape.

> ```python
> shape_parameters = [float(p) for p in value.split(",")]
> ```

Indeed this is cleaner, however I think some additional manipulations are needed, because `value` also contains the name of the shape. For example `value` may be `Drag(5, 0.02)`, so we would first need to extract `5, 0.02` (using regex or some other string gymnastics) and then split.

Again, given that this will be irrelevant pretty soon, I would leave as it is, and merge if someone else confirms it fixes the issue in the qibocal discussion.